File: src/main/scala/searching/BinarySearch.scala
package searching

object BinarySearch extends App {
  val list = List(2, 3, 4, 5, 6, 7, 8, 9)
  println(s"calling binarySearchNormal -- element is present at index ${binarySearchNormal(list, 3)()}")
  println(s"calling binarySearchFunctional -- element is present at index ${binarySearchFunctional(list, 3)}")


  def binarySearchNormal(xs: List[Int], target: Int)(start: Int = 0, end: Int = xs.size - 1): Int = {
    if (start > end) return -1;
    val mid = start + (end - start + 1) / 2
    if (xs(mid) == target)
      return mid
    else if (xs(mid) > target)
      return binarySearchNormal(xs, target)(start, mid - 1)
    else (xs(mid) < target)
    binarySearchNormal(xs, target)(mid + 1, end)

  }

  def binarySearchFunctional(xs: List[Int], target: Int): Int = {
    def bsf(xs: List[Int], target: Int)(start: Int, end: Int): Int = {
      if (start > end) return -1;
      val mid = start + (end - start + 1) / 2
      xs match {
        case (arr: List[Int]) if (arr(mid) == target) => mid
        case (arr: List[Int]) if (arr(mid) > target) => bsf(xs, target)(start, mid - 1)
        case (arr: List[Int]) if (arr(mid) < target) => bsf(xs, target)(mid + 1, end)
      }

    }

    bsf(xs, target)(0, xs.size - 1)
  }

}

File: src/main/java/search/BinarySearch.java
package search;

public class BinarySearch {
  public static int findIterative(int[] array, int key) {
    int start = 0;
    int end = array.length - 1;
    while (start <= end) {
      int mid = (start + end) / 2;
      if (key < array[mid]) end = mid - 1;
      else if (key > array[mid]) start = mid + 1;
      else return mid;
    }
    return -1;
  }

  public static int findRecursive(int[] array, int start, int end, int key) {
    if (start <= end) {
      int mid = (start + end) / 2;
      if (key < array[mid]) return findRecursive(array, start, mid - 1, key);
      else if (key > array[mid]) return findRecursive(array, mid + 1, end, key);
      else return mid;
    }
    return -1;
  }
}

